File: altgraph_web_app/Services/Repositories/TripleRepository.cs
using altgraph_web_app.Models;
using Microsoft.Azure.CosmosRepository;

namespace altgraph_web_app.Services.Repositories
{
  public class TripleRepository
  {
    public IRepository<Triple> Triples { get; set; }
    public TripleRepository(IRepository<Triple> triples)
    {
      Triples = triples;
    }
    public async Task<long> CountAllDocuments()
    {
      return await Triples.CountAsync();
    }

    public async Task<long> CountAllTriples()
    {
      return await Triples.CountAsync(x => x.Doctype == "triple");
    }

    public async Task<long> CountAllLibraries()
    {
      return await Triples.CountAsync(x => x.Doctype == "library");
    }

    public async Task<long> CountAllAuthors()
    {
      return await Triples.CountAsync(x => x.Doctype == "author");
    }

    public async Task<long> CountAllMaintainers()
    {
      return await Triples.CountAsync(x => x.Doctype == "maintainer");
    }

    public async Task<long> GetNumberOfDocsWithSubjectLabel(string subjectLabel)
    {
      return await Triples.CountAsync(x => x.SubjectLabel == subjectLabel);
    }

    public async Task<IEnumerable<Triple>> GetByDoctype(string doctype)
    {
      return await Triples.GetAsync(x => x.Doctype == doctype);
    }

    public async Task<IEnumerable<Triple>> GetByPkLobAndSubjects(string pk,      // "pk": "triple|123"
            string lob,
            string subjectType,
            string objectType)
    {
      return await Triples.GetAsync(x => x.Pk == pk && x.Lob == lob && x.SubjectType == subjectType && x.ObjectType == objectType);
    }


    public async Task<IEnumerable<Triple>> GetByTenantAndSubjectLabels(
            string tenant,
             List<String> subjectLabels)
    {
      return await Triples.GetAsync(x => x.Tenant == tenant && subjectLabels.Contains(x.SubjectLabel));
    }

    public async Task<IEnumerable<Triple>> GetByPkTenantAndSubjectTypes(
            string pk,
            string tenant,
            List<String> subjectTypes)
    {
      return await Triples.GetAsync(x => x.Pk == pk && x.Tenant == tenant && subjectTypes.Contains(x.SubjectType));
    }

    public async Task<IEnumerable<Triple>> FindByTenantAndLobAndSubjectLabelsIn(string tenant, string lob, List<string> values)
    {
      string pk = "triple|" + tenant;
      return await Triples.GetAsync(x => x.Pk == pk && values.Contains(x.SubjectLabel));
    }
  }
}

File: java/altgraph_web_app/src/main/java/org/cjoakim/cosmos/altgraph/data/common/repository/TripleRepository.java
package org.cjoakim.cosmos.altgraph.data.common.repository;

import com.azure.spring.data.cosmos.repository.CosmosRepository;
import com.azure.spring.data.cosmos.repository.Query;
import org.cjoakim.cosmos.altgraph.data.common.model.npm.Triple;
import org.springframework.data.repository.query.Param;
import org.springframework.stereotype.Component;
import org.springframework.stereotype.Repository;

import java.util.List;

/**
 * This class is a Spring Data Repository for Cosmos DB NPM Triple documents.
 * Chris Joakim, Microsoft, November 2022
 */

@Component
@Repository
public interface TripleRepository extends CosmosRepository<Triple, String> { // }, TripleRepositoryExtensions {

    Iterable<Triple> findBySubjectType(String subjectType);

    Iterable<Triple> findBySubjectLabel(String subjectLabel);

    Iterable<Triple> findByTenantAndSubjectLabel(String tenant, String subjectLabel);

    @Query("select value count(1) from c")
    long countAllDocuments();

    @Query("select value count(1) from c where c.doctype = 'triple'")
    long countAllTriples();

    @Query("select value count(1) from c where c.doctype = 'library'")
    long countAllLibraries();

    @Query("select value count(1) from c where c.doctype = 'author'")
    long countAllAuthors();

    @Query("select value count(1) from c where c.doctype = 'maintainer'")
    long countAllMaintainers();

    @Query("select value count(1) from c where c.subjectLabel = @subjectLabel")
    long getNumberOfDocsWithSubjectLabel(@Param("subjectLabel") String subjectLabel);

    @Query("select * from c where c.doctype = @doctype")
    List<Triple> getByDoctype(@Param("doctype") String doctype);

    // select * from c where c.pk = 'triple|123' and c.lob = 'npm' and c.subjectType = 'library' and c.objectType = 'library'
    @Query("select * from c where c.pk = @pk and c.lob = @lob and c.subjectType = @subjectType and c.objectType = @objectType")
    List<Triple> getByPkLobAndSubjects(
            @Param("pk") String pk,      // "pk": "triple|123"
            @Param("lob") String lob,
            @Param("subjectType") String subjectType,
            @Param("objectType") String objectType);

    @Query("select * from c where c.tenant = @tenant and ARRAY_CONTAINS(@subjects, c.subjectLabel)")
    List<Triple> getByTenantAndSubjectLabels(
            @Param("tenant") String tenant,
            @Param("subjects") List<String> subjectLabels);

    @Query("select * from c where c.pk = @pk and c.doctype = 'triple' and c.tenant = @tenant and ARRAY_CONTAINS(@subjectTypes, c.subjectType)")
    List<Triple> getByPkTenantAndSubjectTypes(
            @Param("pk") String pk,
            @Param("tenant") String tenant,
            @Param("subjectTypes") List<String> subjectTypes);
}
